File: utils/NoiseGenerator.js
/**
 * Noise Generator Class
 * Implements Perlin-like noise for procedural generation
 */
class NoiseGenerator {
    constructor(seed = Date.now()) {
        this.seed = seed;
        this.permutation = [];
        this.initPermutation();
    }
    
    /**
     * Initialize permutation table
     */
    initPermutation() {
        // Initialize with values 0-255
        const p = [];
        for (let i = 0; i < 256; i++) {
            p[i] = i;
        }
        
        // Shuffle using seed
        let rng = this.seed;
        for (let i = 255; i > 0; i--) {
            // Simple LCG random number generator
            rng = (1103515245 * rng + 12345) & 0x7fffffff;
            const j = rng % (i + 1);
            [p[i], p[j]] = [p[j], p[i]];
        }
        
        // Duplicate for overflow handling
        this.permutation = [...p, ...p];
    }
    
    /**
     * Fade function for smooth interpolation
     */
    fade(t) {
        return t * t * t * (t * (t * 6 - 15) + 10);
    }
    
    /**
     * Linear interpolation
     */
    lerp(t, a, b) {
        return a + t * (b - a);
    }
    
    /**
     * Gradient function
     */
    grad(hash, x, y) {
        const h = hash & 15;
        const u = h < 8 ? x : y;
        const v = h < 4 ? y : (h === 12 || h === 14 ? x : 0);
        return ((h & 1) === 0 ? u : -u) + ((h & 2) === 0 ? v : -v);
    }
    
    /**
     * Generate 2D noise value (0.0 to 1.0)
     */
    noise2D(x, y) {
        // Find unit square
        const X = Math.floor(x) & 255;
        const Y = Math.floor(y) & 255;
        
        // Relative x, y in square
        x -= Math.floor(x);
        y -= Math.floor(y);
        
        // Fade curves
        const u = this.fade(x);
        const v = this.fade(y);
        
        // Hash coordinates
        const p = this.permutation;
        const aa = p[p[X] + Y];
        const ab = p[p[X] + Y + 1];
        const ba = p[p[X + 1] + Y];
        const bb = p[p[X + 1] + Y + 1];
        
        // Blend results
        const result = this.lerp(v,
            this.lerp(u, this.grad(aa, x, y), this.grad(ba, x - 1, y)),
            this.lerp(u, this.grad(ab, x, y - 1), this.grad(bb, x - 1, y - 1))
        );
        
        // Map from [-1, 1] to [0, 1]
        return (result + 1.0) * 0.5;
    }
    
    /**
     * Generate fractal noise (multiple octaves)
     */
    fractalNoise2D(x, y, octaves = 4, persistence = 0.5) {
        let total = 0;
        let frequency = 1;
        let amplitude = 1;
        let maxValue = 0;
        
        for (let i = 0; i < octaves; i++) {
            total += this.noise2D(x * frequency, y * frequency) * amplitude;
            maxValue += amplitude;
            amplitude *= persistence;
            frequency *= 2;
        }
        
        return total / maxValue;
    }
    
    /**
     * Generate noise in specific range
     */
    noiseRange(x, y, min, max) {
        const n = this.noise2D(x, y);
        return min + n * (max - min);
    }
    
    /**
     * Set new seed
     */
    setSeed(seed) {
        this.seed = seed;
        this.initPermutation();
    }
}

File: cpp/src/utils/NoiseGenerator.cpp
#include "utils/NoiseGenerator.h"
#include <cmath>
#include <algorithm>

NoiseGenerator::NoiseGenerator(uint32_t seed)
    : seed(seed)
{
    initPermutation();
}

void NoiseGenerator::setSeed(uint32_t newSeed) {
    seed = newSeed;
    initPermutation();
}

void NoiseGenerator::initPermutation() {
    // Initialize permutation table with values 0-255
    permutation.resize(512);
    
    // Fill first 256 values
    for (int i = 0; i < 256; ++i) {
        permutation[i] = i;
    }
    
    // Shuffle using seed
    uint32_t rng = seed;
    for (int i = 255; i > 0; --i) {
        // Simple LCG random number generator
        rng = (1103515245 * rng + 12345) & 0x7fffffff;
        int j = rng % (i + 1);
        std::swap(permutation[i], permutation[j]);
    }
    
    // Duplicate for overflow handling
    for (int i = 0; i < 256; ++i) {
        permutation[256 + i] = permutation[i];
    }
}

float NoiseGenerator::fade(float t) const {
    // Smoothstep function: 6t^5 - 15t^4 + 10t^3
    return t * t * t * (t * (t * 6 - 15) + 10);
}

float NoiseGenerator::lerp(float t, float a, float b) const {
    return a + t * (b - a);
}

float NoiseGenerator::grad(int hash, float x, float y) const {
    // Convert low 4 bits of hash into gradient direction
    int h = hash & 15;
    float u = h < 8 ? x : y;
    float v = h < 4 ? y : h == 12 || h == 14 ? x : 0;
    return ((h & 1) == 0 ? u : -u) + ((h & 2) == 0 ? v : -v);
}

float NoiseGenerator::noise2D(float x, float y) const {
    // Find unit square that contains point
    int X = static_cast<int>(std::floor(x)) & 255;
    int Y = static_cast<int>(std::floor(y)) & 255;
    
    // Find relative x, y in square
    x -= std::floor(x);
    y -= std::floor(y);
    
    // Compute fade curves
    float u = fade(x);
    float v = fade(y);
    
    // Hash coordinates of square corners
    int aa = permutation[permutation[X] + Y];
    int ab = permutation[permutation[X] + Y + 1];
    int ba = permutation[permutation[X + 1] + Y];
    int bb = permutation[permutation[X + 1] + Y + 1];
    
    // Blend results from corners
    float result = lerp(v,
        lerp(u, grad(aa, x, y), grad(ba, x - 1, y)),
        lerp(u, grad(ab, x, y - 1), grad(bb, x - 1, y - 1))
    );
    
    // Map from [-1, 1] to [0, 1]
    return (result + 1.0f) * 0.5f;
}

float NoiseGenerator::fractalNoise2D(float x, float y, int octaves, float persistence) const {
    float total = 0.0f;
    float frequency = 1.0f;
    float amplitude = 1.0f;
    float maxValue = 0.0f;
    
    for (int i = 0; i < octaves; ++i) {
        total += noise2D(x * frequency, y * frequency) * amplitude;
        maxValue += amplitude;
        amplitude *= persistence;
        frequency *= 2.0f;
    }
    
    return total / maxValue;
}

float NoiseGenerator::noiseRange(float x, float y, float min, float max) const {
    float n = noise2D(x, y);
    return min + n * (max - min);
}
